lib/daru/category.rb

```diff
@@ -0,0 +1,809 @@
+    # @example
+    #   dv = Daru::Vector.new [:a, 1, :a, 1, :c], type: :category
+    #   dv.add_category :b
+    #   dv[0] = :b
+    #   dv
+    #   # => #<Daru::Vector(5)>
+    #   #   0   b
+    #   #   1   1
+    #   #   2   a
+    #   #   3   1
+    #   #   4   c
+    def []= *indexes, val
+      positions = @index.pos(*indexes)
+
+      if positions.is_a? Numeric
+        modify_category_at positions, val
+      else
+        positions.each { |pos| modify_category_at pos, val }
+      end
+      self
+    end
+
+    # Modifies values at specified positions.
+    # @param [Array] positions positions at which to modify value
+    # @param [object] val value to assign at specific positions
+    # @return modified vector
+    # @example
+    #   dv = Daru::Vector.new [:a, 1, :a, 1, :c], type: :category
+    #   dv.add_category :b
+    #   dv.set_at [0, 1], :b
+    #   # => #<Daru::Vector(5)>
+    #   #   0   b
+    #   #   1   b
+    #   #   2   a
+    #   #   3   1
+    #   #   4   c
+    def set_at positions, val
+      validate_positions(*positions)
+      positions.map { |pos| modify_category_at pos, val }
+      self
+    end
+
+    # Size of categorical data.
+    # @return total number of values in the vector
+    # @example
+    #   dv = Daru::Vector.new [:a, 1, :a, 1, :c], type: :category
+    #   dv.size
+    #   # => 5
+    def size
+      @array.size
+    end
+
+    # Tells whether vector is ordered or not.
+    # @return [Boolean] true if vector is ordered, false otherwise
+    # @example
+    #   dv = Daru::Vector.new [:a, 1, :a, 1, :c], type: :category
+    #   dv.ordered?
+    #   # => false
+    def ordered?
+      @ordered
+    end
+
+    # Make categorical data ordered or unordered.
+    # @param [Boolean] bool true if categorical data is to be to ordered, false otherwise
+    # @example
+    #   dv = Daru::Vector.new [:a, 1, :a, 1, :c], type: :category
+    #   dv.ordered = true
+    #   dv.ordered?
+    #   # => true
+    def ordered= bool
+      @ordered = bool
+    end
+
+    # Returns all the categories with the inherent order
+    # @return [Array] categories of the vector with the order
+    # @example
+    #   dv = Daru::Vector.new [:a, 1, :a, 1, :c],
+    #     type: :category,
+    #     categories: [:a, :b, :c, 1]
+    #   dv.categories
+    #   # => [:a, :b, :c, 1]
+    def categories
+      @cat_hash.keys
+    end
+
+    alias_method :order, :categories
+
+    # Sets order of the categories.
+    # @note If extra categories are specified, they get added too.
+    # @param [Array] cat_with_order categories specifying their order
+    # @example
+    #   dv = Daru::Vector.new [:a, 1, :a, 1, :c], type: :category
+    #   dv.categories = [:a, :b, :c, 1]
+    #   dv.categories
+    #   # => [:a, :b, :c, 1]
+    def categories= cat_with_order
+      validate_categories(cat_with_order)
+      add_extra_categories(cat_with_order - categories)
+      order_with cat_with_order
+      self
+    end
```

Comment:
It makes sense, in fact. In many (questionable, yet useful) cases:
```ruby
if (x.foo = some_call)
  puts x
end
# or
x.foo = some_call or raise RuntimeError
# or
y.bar = x.foo = some_call
```
...you'd prefer the result of `x.foo = some_call` to be "intuitive" (e.g. the same as `x = y`), not depending on `#foo=` implementation.